src/api/app/controllers/webui/badge_controller.rb

```diff
@@ -0,0 +1,139 @@
+class Webui::BadgeController < Webui::WebuiController
+  before_action :set_project, only: [:show]
+  before_action :require_package, only: [:show]
+  before_action :require_login, except: [:show]
+
+  def show
+    # Badge height
+    @height = 20
+    # We create an array that holds all the individual segments of the badge,
+    # this way in the future adding more badge variants will be easier.
+    @badge = []
+    # The possible fields inside of a badge hash are:
+    # icon: it's provided as Rails asset, to make it easier to get content type and hash
+    # text: the text that will be shown in that segment
+    # width: text_width function is commonly used here, but we can provide other
+    #        values here, to add additional padding or draw bar graphs with just
+    #        colors without text
+    # color: background color, we currently have the following options:
+    #        [:gray, :green, :yellow, :orange, :blue, :red]
+    info = { text: 'build service', color: color(:gray), font_size: 11, icon_size: 14 }
+    icon = Rails.application.assets['obs-logo_no-text.svg']
+    info[:icon] = "data:#{icon.content_type};base64,#{Base64.strict_encode64(icon.source)}"
+    info[:text_width] = (text_width('build service') * info[:font_size]).ceil
+    @badge << info
+    project = Project.find_by_name(params[:project])
+    package = Package.find_by_project_and_name(params[:project], params[:package])
+    # Return if the following part will fail
+    return badge_processing unless package.buildresult(project).results[package.name]
+
+    results = simplify_buildresults(package.buildresult(package.project).results[package.name])
+    status = results.group_by(&:code)
+    result = { color: :blue, text: status.key(status.values.max_by(&:length)) } # set this by default as it's the intermediate state
+    setup_type(result, results.length, status)
+    result[:font_size] = 11
+    result[:text_width] = (text_width(result[:text]) * result[:font_size]).ceil
+    @badge << result
+    badge_processing
+  end
+
+  private
+
+  def color(value)
+    case value
+    when :green then '#97ca00'
+    when :yellow then '#dfb317'
+    when :orange then '#fe7d37'
+    when :blue then '#007ec6'
```

Comment:
This can be replaced by constants:
```
GREEN = '#97ca00'
YELLOW = '#dfb317'
```

Also, this looks like view code, not controller code...